model/riscv_insts_zicsr.sail

```diff
@@ -9,23 +9,25 @@
 /* ****************************************************************** */
 /* This file specifies the instructions in the 'Zicsr' extension.     */
 /* ****************************************************************** */
-union clause ast = CSR  : (csreg, regidx, regidx, bool, csrop)
+union clause ast = CSRReg  : (csreg, regidx, regidx, csrop)
+union clause ast = CSRImm  : (csreg, bits(5), regidx, csrop)
 
 mapping encdec_csrop : csrop <-> bits(2) = {
   CSRRW <-> 0b01,
   CSRRS <-> 0b10,
   CSRRC <-> 0b11
 }
 
-mapping clause encdec = CSR(csr, rs1, rd, is_imm, op)
-  <-> csr @ encdec_reg(rs1) @ bool_bits(is_imm) @ encdec_csrop(op) @ encdec_reg(rd) @ 0b1110011
+mapping clause encdec = CSRReg(csr, rs1, rd, op)
+  <-> csr @ encdec_reg(rs1) @ 0b0 @ encdec_csrop(op) @ encdec_reg(rd) @ 0b1110011
 
-function clause execute CSR(csr, rs1, rd, is_imm, op) = {
-  let regidx(rs1_bits) = rs1;
-  let rs1_val : xlenbits = if is_imm then zero_extend(rs1_bits) else X(rs1);
+mapping clause encdec = CSRImm(csr, imm, rd, op)
+  <-> csr @ imm @ 0b1 @ encdec_csrop(op) @ encdec_reg(rd) @ 0b1110011
+
+function doCSR(csr : csreg, rs1_val : xlenbits, rd : regidx, op : csrop) -> Retired = {
   let isWrite : bool = match op {
     CSRRW  => true,
-    _      => if is_imm then unsigned(rs1_val) != 0 else rs1 != zreg
+    _      => unsigned(rs1_val) != 0
```

Comment:
I'm not sure this is correct, looking at https://github.com/riscv/riscv-isa-manual/blob/main/src/zicsr.adoc it only says write = false for imm=0 or rs1=0, but if rs1 happens to hold a zero value it still performs the write.

> Note that if rs1 specifies a register other than x0, and that register holds a zero value, the instruction will not action any attendant per-field side effects, but will action any side effects caused by writing to the entire CSR.

Reply:
Oh, hah, whoops; I was a victim of the kind of type confusion I was hoping to prevent by making things more explicit.

I'll pass `isWrite` to the `doCSR` and have the case arms do the appropriate discrimination.